stix_shifter_modules/nozomi/stix_transmission/connector.py

```diff
@@ -0,0 +1,217 @@
+                        jwt_token = await self.get_token()
+                        token_generated = True
+                        continue
+                    else:
+                        error = response_dict.get('error', '')
+                        if response_wrapper.code == 403:
+                            error = 'Query length is too long or Invalid Query'
+                        return_obj = self.exception_response(response_wrapper.code, error)
+                        data = []
+                        break
+
+                if data:
+                    data = self.get_results_data(data)
+                    if metadata:
+                        return_obj['data'] = data if data else []
+                    else:
+                        return_obj['data'] = data[int(offset): total_records] if data else []
+                    if jwt_token and next_url and result_count < self.api_client.result_limit:
+                        return_obj['metadata'] = {"result_count": result_count, "jwtToken": jwt_token,
+                                                  "next_url": next_url}
+                else:
+                    if not return_obj.get('error') and return_obj.get('success') is not False:
+                        return_obj['success'] = True
+                        return_obj['data'] = []
+            else:
+                return_obj['success'] = True
+                return_obj['data'] = []
+
+        except InvalidMetadataException as ex:
+            return_obj = self.exception_response(422, f'Invalid metadata: {str(ex)}')
+
+        except Exception as ex:
+            return_obj = self.exception_response(None, str(ex))
+        return return_obj
+
+    async def ping_connection(self):
+        """
+        Ping the endpoint
+        :return: return_object, dict
+        """
+        return_obj = {}
+        try:
+            token = await self.get_token()
+            response = await self.api_client.ping_data_source(token)
+            response_code = response.code
+            response_dict = json.loads(response.read().decode('utf-8'))
+            if response_code == 200:
+                return_obj['success'] = True
+            else:
+                return_obj = self.exception_response(response_code, response_dict.get('errorSummary', ''))
+        except Exception as ex:
+            return_obj = self.exception_response(None, str(ex))
+        return return_obj
+
+    async def get_token(self):
+        """ Generate new token"""
+        response = await self.api_client.generate_token()
+        response_code = response.code
+        response_txt = response.read().decode('utf-8')
+        response_json = json.loads(response_txt)
+        if response_code == 200:
+            response_header = response.headers
+            if 'Authorization' in response_header:
+                token = response_header.get('Authorization')
+            return token
+        raise Exception(response_json['errors'])
```

Comment:
If the exception isn't 200, it should give back the error code and the message associated with it. It should probably go through the "exception_response()" method.

Reply:
If we use the exception_response function in the failure case, it will return an error message as a dict.
While calling the function, we need to check if this function returns a token or an error message.
token generation is called twice in the same function, one for generating new tokens and another for regenerating expired tokens.
So in both places, we need to add a check this function returns a token or error message.
To avoid the checks, we are raising the error, so it will come in the exception block and the error will be handled.

Reply:
Can you change the Exception to a custom Exception object (something like TokenException() maybe) that accepts both an error message and an http status code? Then you return that object to the calling method (ping or connect) where it catches that specific TokenException and handles it there? 

That would alleviate my main concern that all of the non-200 responses should be handled the same way, regardless of where they come from.

Reply:
Instead of raising the error, handle_api_exception function is used for handling the non 200 response.